Implement periodic splines in ``make_splrep``

czgdp1807 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
<!--Please explain your changes.-->

I have implemented periodic splines in ``make_splrep``. I have divided the problem in two parts,

1. Figuring out the knots. This is done in the following steps,

- Create the data matrices - R , H1 , H2 and CORR , each having m rows. R will contain normal evaluation of b-splines. H1 and H2 will correspond to h1 and h2 . They will be zeros if l <= n - 2k (because that would mean correction is not present in that row). CORR would contain l > n - 2k to denote which data point have the correction part.
- Now we will use H1 and H2 and CORR to apply givens rotation to R . We will also generate R2 (the QR decomposition of correction matrix).
- With R and R2 we will do backward substitution to generate the coefficients.
- Then the last k coefficients will be filled using the first k .

2. Figuring out the discontinuity matrix and then solving `f(p) = s` for `p`. This is done in the following steps,

- Compute, fp0 and fp (for periodic case the method is different, we don’t use knots, only for fp0 , for fp we are good).
- Compute discontinuity matrix, with corrections.
- Create augmented matrix with 1/p applied to discontinuity matrix.
- Do backsubstiution.
- Compute residuals.
- Update p. Go to step 3.

#### Additional information
<!--Any additional information you think is important.-->

I have also added tests as well.

I know this is not perfect and I have taken most of the logic from Fortran code. I have split it into different understandable parts and then re-implemented in C++. 

I am open to feedback. Simplifying the mathematics which will lead to further simplification of the C++ code I have written is what I am looking forward to.

czgdp1807 commented:
@ev-br Would love if you go through the description and review my changes. I am looking forward to further simplification of the mathematics and subsequently the C++ code.

Thank you.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -1367,7 +1367,7 @@ def _make_periodic_spline(x, y, t, k, axis):
 
 def make_interp_spline(x, y, k=3, t=None, bc_type=None, axis=0,
-                       check_finite=True):
+                       check_finite=True, periodic=False):
```
Please reuse the already-existing `bc_type={"periodic", "not-a-knot"}` API. This relates to all public functions: make_splrep, make_splprep (is it a part of this first PR?) and generate_knots. There is machinery for validating the `bc_type` argument in `_bsplines.py`, feel free to reuse. No need to worry about other values of bc_type: just literals, "not-a-knot" (default) and "periodic". 

Private functions can of course keep using whatever is convenient, `periodic={True,False}`.

ev-br commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -3564,6 +3564,32 @@ def test_s0_vs_not(self):
+
+        y = np.exp(x)
+        tck = make_splrep(x, y, s=1e-8, periodic=True)
+        xp_assert_close(np.max(np.abs(splev(x, tck) - y)), 534.4916122179081)
```
Not sure what is this. I'd expect that there's a check and if the data is not periodic, it just bails out with a ValueError.

ev-br commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -3564,6 +3564,32 @@ def test_s0_vs_not(self):
+
+        y = np.cos(x)
+        tck = make_splrep(x, y, s=1e-8, periodic=True)
+        xp_assert_close(splev(x, tck), y, atol=1e-5, rtol=1e-4)
```
Minor, trivial: can use `tck(x)` (rename from tck to spl or something).

ev-br commented on scipy/interpolate/tests/test_bsplines.py:
```diff
@@ -3564,6 +3564,32 @@ def test_s0_vs_not(self):
+
+        y = 5*np.sin(x) + np.cos(x)*3
+        tck = make_splrep(x, y, s=1e-8, periodic=True)
+        xp_assert_close(splev(x, tck), y, atol=1e-5, rtol=1e-4)
```
Overall this test looks nice but leaves a question of why s is always this small. I'd add some more tests to probe the cases of 1) no internal knots, just the initial estimate, 2) some internal knots, not all data, 3) smaller s, more knots, finally, all data --- then the result should agree with CubicSpline to a tight tolerance, I'd expect.

Also: you only test k=3. Does splrep(x, y, periodic=True) handle k=1 and/or k=2? If yes, these should be tested (careful with k=2: for interpolation generate_knots switches from data points to Greville abscissae); if not, these should raise fast. [if it's hard to implement, it's fine to bail out and leave a comment that it's not implemented and what is hard].

ev-br reviewed (changes requested):
Looks promising!

This is an initial cursory review, am only looking at the public API and tests. The PR is large, so I won't be able to make it in a single review.

It'd be really helpful to have code comments with references to specific equations of the original papers + the Fortran code you are translating.

I'm not sure about additional H1, H2 arrays, will be able to say more when I study the code some more. References to the original papers and the original implementation would be of great help.

czgdp1807 commented:
I have adressed https://github.com/czgdp1807/scipy/pull/19#discussion_r2029432093, https://github.com/czgdp1807/scipy/pull/19#discussion_r2029433480 and https://github.com/czgdp1807/scipy/pull/19#discussion_r2029434372. 

Now I will first focus on testing my work. Alongside I will update the code with comments and links to Fortran code wherever necessary.

czgdp1807 commented:
I found a bug in periodic splines for constant functions. Will work on fixing it and uncomment the test.

czgdp1807 commented:
Fixed the bug. It was a corner case, we can exit early if `fpms < acc` already. It would mean constant function is the solution to our problem, no fitting needed. 

Except, `Fperiodic`, everything is tested for `make_splrep(..., periodic=True)` with the same test cases as non-periodic splines.

ev-br commented:
> Tested for more degrees (k). So we are good on testing front.

Not quite there yet, by the looks of it :-(.

czgdp1807 commented:
I see. No worries. Will use more cases from non-periodic splines and apply it on periodic case. Will keep updating here.

czgdp1807 commented:
Added all the possible tests for periodic splines. Now I will see how the CI responds. Meanwhile I will add comments wherever possible. Then I will fix the bugs coming in CI.

czgdp1807 commented:
I am able to reproduce the 32-bit failure using Docker on my laptop (followed the exact same instructions as on CI). However, seems like the segmentation fault only in `pytest` mode, when we run, 

```zsh
docker run -v $(pwd):/scipy --platform=linux/i386 quay.io/pypa/manylinux2014_i686 /bin/bash -c "cd /scipy && \
        uname -a && \
        python3.11 -m venv test && \
        source test/bin/activate && \
        python dev.py test -t scipy/interpolate/tests/test_bsplines.py"
```

However, when we run the following example (in isolation),

```python
import numpy as np
from scipy._lib._array_api import xp_assert_close
from scipy.interpolate import generate_knots, make_splrep

def _get_xykt():
    m = 10
    a, b = 0, 2*np.pi
    x = np.linspace(a, b, m)    # nodes
    y  = np.sin(x)
    k = [1, 2, 3, 4, 5, 6]
    s = 1.7e-4

    return x, y, k, s, []

def _test_with_knots(x, y, k, s):
    t = list(generate_knots(x, y, k=k, s=s, periodic=True))[-1]

    spl_auto = make_splrep(x, y, k=k, s=s, periodic=True)
    spl_t = make_splrep(x, y, t=t, k=k, s=s, periodic=True)

    xp_assert_close(spl_auto.t, spl_t.t, atol=1e-15)
    xp_assert_close(spl_auto.c, spl_t.c, atol=1e-15)
    assert spl_auto.k == spl_t.k

def test_with_knots():
    x, y, k, s, _ = _get_xykt()

    for k_ in k:
        _test_with_knots(x, y, k_, s)

test_with_knots()
```

then no errors or segmentation faults. I used the following command,

```zsh
docker run -v $(pwd):/scipy --platform=linux/i386 quay.io/pypa/manylinux2014_i686 /bin/bash -c "cd /scipy && \
        uname -a && \
        python3.11 -m venv test && \
        source test/bin/activate && \
        python dev.py python debug.py"
```

So @rgommers @ev-br what I am thinking is to skip 32-bit tests on CI for periodic splines. May be we can debug it later once our work goes in `main`. IMO, investing time to fix something which might go away after reviews, etc doesn't seem right to me.

What do you folks say?

ev-br commented:
Eventually we'll want it fixed of course. Is OK to punt on this now if you expect large refactorings. I'll do a review pass, hopefully today, possibly tomorrow.

ev-br commented on scipy/interpolate/tests/test_fitpack.py:
```diff
@@ -88,6 +88,15 @@ def err_est(k, d):
                     assert k == 5   # knot length differ in some k=5 cases
+            else:
+                if np.allclose(v[0], v[-1], atol=1e-15):
+                    spl = make_splrep(x, v, k=k, s=s, xb=xb, xe=xe, periodic=per)
```
Minor: This all is in the `if not per` branch, so can remove `periodic=per`

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -89,9 +97,10 @@ def add_knot(x, t, k, residuals):
+def _validate_inputs(x, y, w, k, s, xb, xe, parametric, periodic=False):
     """Common input validations for generate_knots and make_splrep.
     """
+    # TODO: Add validity checks for periodic inputs
```
This TODO would be good to act on earlier than later.

ev-br commented on scipy/interpolate/_bsplines.py:
```diff
@@ -1866,11 +1866,35 @@ def _lsq_solve_qr(x, y, t, k, w):
+        print("periodic.begin")
+        # Ref: https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L221-L238
+        R, H1, H2, offset, nc = _dierckx.data_matrix(x, t, k, w, False, True)
+        assert(y.shape[1] == 1) # TODO: Update QR Reduce to account for y.shape[1] != 1
```
Could you explain what is this?
y.ndim == 2, what is y.shape[1], does this differ between parametric and non-parametric case, or is it something else?

ev-br commented on scipy/interpolate/src/_dierckxmodule.cc:
```diff
@@ -204,6 +262,247 @@ py_qr_reduce(PyObject* self, PyObject *args, PyObject *kwargs)
+        return NULL;
+    }
+
+    k = py_k;
```
I was going to say it's clearly wrong, but no, just confusing: `py_k` is a Py_ssize_t variable. Please keep the naming consistent: `py_something` is a `PyObject*`, `a_something` is a `PyArrayObject *`, and one-letter variables are PODs.

ev-br commented on scipy/interpolate/src/_dierckxmodule.cc:
```diff
@@ -204,6 +262,247 @@ py_qr_reduce(PyObject* self, PyObject *args, PyObject *kwargs)
+
+    if(!PyArg_ParseTupleAndKeywords(args, kwargs, "OOOOnOnn|pp", const_cast<char **>(kwlist),
+                                    &py_a, &py_h1, &py_h2, &py_offs, &nc, &py_y, &py_k, &len_t,
+                                    &init_p, &get_fp)) {
```
The fact that you need to communicate `get_fp` from C to python is a clear sign that something is not right. Ditto for `init_p`.
Not yet sure what.

ev-br reviewed (commented):
Several minor comments inline.

Bigger picture: Let's take a step back. Could you please add an overview of the computation:

1. how you partition the data matrix (incl sizes of various parts)
2. what are h1 and h2, what are their sizes
3. References to specific equations from the paper would be really helpful.

czgdp1807 commented:
I am thinking it is because of a very fine detail in the combination of (32-bit system, python 3.11, numpy 1.25.2 built with scipy_openblas, pytest). This combination is only available on docker. Its hard to debug there. I have to re-run the tests on docker every time a make change. If I can get access to a 32-bit system where I can repro this bug, then I will be able to fix it quickly.

Meanwhile, please feel free to review. I will skip the failing test for now, but I rest assured, once I get approving reviews, I will put my efforts into fixing this bug in the docker environment itself.

czgdp1807 commented:
So I have addressed your review, now coming on to the code flow (step-by-step flow of calls),

`make_splrep(..., bc_type='periodic')` - This is the call to public API.

`_validate_inputs(..., periodic=True)` - It has only one specific call for periodic splines. It makes sure that first and last of `y` vector match. If not, we exit with an error message.

`_make_splrep_impl(..., periodic=True)` - This is where the real computation happens. After setting, `nest` (exactly as per Fortran code),  we can break it down to two parts,

a. `_generate_knots_impl(..., periodic=True)` - So here we set, `nmin`, `nmax`, and `nest` (if its not already set), exactly as per Fortran code. The first noticeable call is `_dierckx.get_residual_p0`. This computes `fp` for constant function. It is the same computation as in [this snippet](https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L107-L123). Once we have `fp` for constant function, for periodic case, we check whether constant function already satisfies the accuracy criteria.  If so, we return the knots for this constant function. (exactly as per [this Fortran code](https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L600-L610)).
If constant function is not the solution, then we initialise the knots as per [this Fortran code](https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L131).
Now we start the iterations to figure out the number and position of knots. At the beginning of each iteration we apply periodicity condition on knots (see, Eq 24 in [1], or [this Fortran code](https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L147-L158)).
The next step is to obtain `fp` for the current set knots (found above). We do it by calling `_get_residuals` which further calls `_lsq_solve_qr(..., periodic=True)` 

Details of `_lsq_solve_qr(..., periodic=True)` are as follows,

i. `_dierckx.data_matrix` - This computes data matrices for periodic case. It gives three matrices and one vector as output. `R`, `H1`, `H2` and `offset`. `R` is simply the evaluation of periodic spline for the current set of knots. `R[i]` corresponds to evaluation of spline for `x[i]`. `H1[i]` and `H2[i]` is the split up of `R[i]`. This split is done exactly as per [this Fortran code](https://github.com/scipy/scipy/blob/10f63b25d1e040cca3d7319dc2edff0c31ef8b7a/scipy/interpolate/fitpack/fpperi.f#L221-L238). In Fortran code you can see, `h1` and `h2` (which are vectors). `H1[i]` and `H2[i]` correspond to `h1` and `h2` for each `x[i]`.
ii. `_dierckx.qr_reduce_periodic` - The data matrices and offset vector obtained from `_dierckx.data_matrix` are used here. We rotate `H1[i]` and `H2[i]` if `offset[i] > n - 4k - 1` and otherwise we rotate `R[i]`. We fill up, `A1` (shape - `len_t - k - 1, k + 1`), `A2` (shape - `len_t - 2*k - 1, k`), `Z` (shape - `len_t - k - 1`) matrices (initialised to 0.0) in the process (iterating through each row of `H1`, `H2` and `R`). `Z` contains the transformations to right hand side i.e., `y`. Here we also compute `fp` for the current set of knots. `fp = sum(yi**2)` where `yi` contains the value after givens transformation to the right hand side is done. We return it if `get_fp = True`. I had to do it because, there is a difference between `fp` for non-periodic splines (residuals.sum()) and periodic splines.
iii. `_dierckx.fpbacp` - This is exact replica of [fpbacp.f](https://github.com/scipy/scipy/blob/main/scipy/interpolate/fitpack/fpbacp.f). Basically backward substitution for periodic splines.

After `_lsq_solve_qr` rest of the iteration is exactly same as for non-periodic splines.

b. After knots generation we solve for `p`. This can be broken down into the following steps,

i. Getting initial value for `p`. We get it via `_lsq_solve_qr` by passing `solve_for_p=True`. This is communicated to `_dierckx.qr_reduce_periodic` which calculates `p` as per [this Fortran code](https://github.com/scipy/scipy/blob/10f63b25d1e040cca3d7319dc2edff0c31ef8b7a/scipy/interpolate/fitpack/fpperi.f#L407-L427).
ii. Now we get the bounds, `f(p=inf)` via `_get_residuals(..., periodic=True, get_fp=True)`. This is communicated to `_dierckx.qr_reduce_periodic`. The process of getting `fp` is explained in second step of details for `_lsq_solve_qr`. We get lower bound as `f(p=0)`, which is nothing but `fp` for constant function. We get it via, `_dierckx.get_residual_p0` as explained above.
iii. `Fperiodic` - This is analogous to `F` but it is for periodic splines.

`Fperiodic.__init__` - Initialising the augmented matrices. The `A1` and `A2` we get from the last iteration of knots generation, we use these matrices to initialise augmented matrices. This whole step is exactly same as [this Fortran code](https://github.com/scipy/scipy/blob/10f63b25d1e040cca3d7319dc2edff0c31ef8b7a/scipy/interpolate/fitpack/fpperi.f#L441-L493) except one difference (we have `G1`, `G2`, `H1`, `H2` as it is). The difference in our case is we don't multiply `H1` (shape - `len_t - 2*k - 2, k + 2`) and `H2` (shape - `len_t - 2*k - 2, k + 1`) with `pinv`. We do it in the iterative step.

`Fperiodic.__call__(self, p)` - Here the steps are simple,

a. Multiply `H1` and `H2` with `pinv = 1/p`.
b. Initialise first `len_t - 2*k - 1` coefficients with the values in `Z` (obtained in knots generation step).
c. Now do QR reduction of `G1`, with `H1` and `G2` with `H2`.
d. Pass these `G1` and `G2` into `_dierckx.fpbacp` and update the coefficients.
e. Compute the residuals using, `_compute_residuals`.
f. Return the residual obtained.

So this is the whole process. Things where I think we can do better,

1. Getting `fp` from C++ code. May be we can use `residuals` vector itself in Python to compute `fp`. I will have to check. 
2. Computing for initial value of `p` as well. If we can utilise `A1` and `A2` in the Python code itself to compute `p`.

I implemented the above to the best of my abilities (in my understanding of the Fortran code, the Mathematics behind it). Apologies if I did any mistake. I still think we can improve a lot from here. Just need inputs on the Mathematics using which we can simplify things.

ev-br commented:
Thanks Gagan, this is very helpful.

>I implemented the above to the best of my abilities (in my understanding of the Fortran code, the Mathematics behind it). Apologies if I did any mistake. I still think we can improve a lot from here. Just need inputs on the Mathematics using which we can simplify things.

To state the obvious --- my need to review does not mean there's something wrong with the code, in any shape or form. Any code can be improved, and this is just the right time to 1) for me, to familiarize with the code, and 2) consider potential improvements.  Hence a detailed review. No more no less.

And yes, this is a very difficult piece of tricky work.

czgdp1807 commented:
Yes. I am having a strong feeling that we can improve a lot from here. Once you review, we can figure out simplification steps then. At least for now we have something which works on 64-bit systems.

ev-br commented:
Sorry for the delay in reviewing Gagan. While I work through the details, here's the first ask: 

 `_lsq_solve_qr(x, y, t, k, w, periodic=False, solve_for_p=False, get_fp=False)`

Please remove the `get_fp` argument and make it always return `fp`. For this, you'll need to move the `fp` computation for a non-periodic case to C, I believe.

ev-br commented:
You can get the sum of the residuals from `_qr_reduce`. This is what the Fortran code does anyway.

czgdp1807 commented:
I shifted the `residuals` computation to C++ implementations of both `_dierckx.fpback` and `_dierckx.fpbacp`. `fp` is computed for non-periodic case in `_dierckx.fpback`. Should solve the issue for removing `get_fp` argument from `_lsq_solve_qr`.

cc: @ev-br

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -215,51 +231,105 @@ def generate_knots(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None):
+    per = xe - xb
+    # Ref: https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L107-L123
+    # Computes fp0 for constant function
+    fpold = _dierckx.get_residual_p0(y, w)
```
This is only needed if `periodic is True`, thus please move it to inside the `if`.

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -215,51 +231,105 @@ def generate_knots(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None):
+                t[i] = x[0] - (k - i) * per
+                t[i + k + 1] = x[m - 1] + i * per
+            yield t
+            return
```
Do we have test coverage of this line code path? For both min==nmax (IOW, m=k+1?) and for when `fpold -s < acc` but `nmax > nmin`.

czgdp1807 replied:
Actually for `nmax == nmin` we need to have (for periodic case), `m = 2`. This is because in periodic case, `nmin = 2*(k + 1)` and `nmax = m + 2*k`. Solve for `nmin = nmax`, we will get `m = 2`. Do you want me to add a test for `m = 2`. It will only work for, `k = 1`. 

> for when `fpold - s < acc` but `nmax > nmin`.

We have a test function, `TestMakeSplrepPeriodic::test_no_internal_knots` for the above case.

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -215,51 +231,105 @@ def generate_knots(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None):
+        t = np.asarray([xb]*(k+1) + [xe]*(k+1), dtype=float)
+    else:
+        # Ref: https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L131
+        # For periodic splines initial set of knots is all zeros except (k + 2)-th knot
```
Not really: you're making the periodic extension in the loop below. Please mention this in this comment.

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -215,51 +231,105 @@ def generate_knots(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None):
         # construct the LSQ spline with this set of knots
+        # N.B. - Check _lsq_solve_qr which is called
+        # via _get_residuals for logic behind
+        # computation of fp for periodic splines
```
Not sure if this comment is helpful

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -215,51 +231,105 @@ def generate_knots(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None):
+            t[n - k - 1] = xe
+            for j in range(1, k + 1):
+                t[k - j] = t[n - k - j - 1] - per
+                t[n - k + j - 1] = t[k + j] + per
```
Is this not `periodic_knots(x, k)`?

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -300,8 +373,9 @@ def _generate_knots_impl(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None)
-                residuals = _get_residuals(x, y, t, k, w=w)
+                residuals, _ = _get_residuals(x, y, t, k, w=w, periodic=periodic)
 
+        fpold = fp
```
Please check if this is still needed (same is done at the beginning of the loop)

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -501,12 +575,85 @@ def __call__(self, p):
+
+        return fp - self.s
+
+# Solver class for f(p) for periodic splines
```
Please explain the difference to `F`. What's sorely needed is an explanation of A1, A2, Z and H1, H2.
Ideally, have something similar to the docstring of F.

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -501,12 +575,85 @@ def __call__(self, p):
+            (R is None and Y is None)):
+            # Ref: https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L171-L215
+            # The computation in the above link is performed for a
+            # given set of knots i.e., t vector
```
This comment is cryptic. Please expand what you meant.

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -501,12 +575,85 @@ def __call__(self, p):
+            # The computation in the above link is performed for a
+            # given set of knots i.e., t vector
+            (R, A1, A2, Z), _, _, _, _ = _lsq_solve_qr(
+                x, y, t, k, w, periodic=True, solve_for_p=True)
```
This is extremely confusing. This seems to be the only place where you use `solve_for_p=True`, but you do not return it.

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -501,12 +575,85 @@ def __call__(self, p):
+
+        # Ref: https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L441-L493
+        # Note the for H1 and H2, pinv is not multiplied in the initialisation step.
+        # Check __call__ for multiplication of pinv with H1 and H2 respectively
```
This comment is not helpful. Please expand.
EDIT: ah, "pinv" is not a pseudoinverse function, it's `1/p`. So it's the division by `p`. Still cryptic. And the "check that for this" style comments are really not clear. What is the statement are you making? Thtat the multiplication happens in `__call__`? Then say it. That there's something non-trivial happening in `__call__`? Then say it.

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -501,12 +575,85 @@ def __call__(self, p):
+        # Ref: https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L441-L493
+        # Note the for H1 and H2, pinv is not multiplied in the initialisation step.
+        # Check __call__ for multiplication of pinv with H1 and H2 respectively
+        G1, G2, H1, H2, offset = _dierckx.init_augmented_matrices(A1, A2, b, len(t), k)
```
What you add an explanation of A1,A2, also explain what are G1,G2, H1,H2

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -501,12 +575,85 @@ def __call__(self, p):
+
+        # Ref: https://github.com/scipy/scipy/blob/596b586e25e34bd842b575bac134b4d6924c6556/scipy/interpolate/fitpack/fpperi.f#L498-L534
+        _dierckx.qr_reduce_augmented_matrices(
+            G1, G2, H1, H2, c, self.offset_, len(self.t), self.k)
```
From reading the code I've no way of knowing what are "augmented matrices". This should be explained in the docstring.

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -656,47 +804,77 @@ def _make_splrep_impl(x, y, *, w=None, xb=None, xe=None, k=3, s=0, t=None, nest=
+    if not periodic:
+        p = nc / R[:, 0].sum()
+    else:
+        R, A1, A2, Z = R
```
What's going on here? I cannot follow if `p` is defined or not. Please streamline this.

The whole `solve_for_p` argument and value-based number of outputs is extremely confusing, and should be fixed.

ev-br reviewed (changes requested):
Ok, finally. Here's a review of the python side. The C++ side review will follow shortly. 

My main requests are to explain the algorithm and its various moving parts (A1,A2, H1,H2 etc) and and to cut down on branching.

ev-br commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -171,10 +171,97 @@ data_matrix( /* inputs */
+// and
+// Ref: https://github.com/scipy/scipy/blob/10f63b25d1e040cca3d7319dc2edff0c31ef8b7a/scipy/interpolate/fitpack/fpperi.f#L221-L238
+// For given set of knots, computes H1 and H2 for all the data points except the last.
+// Check the above link for semantics of H1 and H2.
```
As mentioned, "check the above link" is not helpful. Please add comments in the style of the `data_matrix`.
Here ISTM, the computations is nearly the same as that in `data_matrix`, with additional something for H1 and H2 for each row. Please say it and explain the difference.

ev-br reviewed (changes requested):
Several requests below. As is, it is sadly neither comprehensible nor mergeable. 
Once there are comments, I'll go through the code again.

czgdp1807 commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -215,51 +231,105 @@ def generate_knots(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None):
 
 
-def _generate_knots_impl(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None):
+def _generate_knots_impl(x, y, *, w=None, xb=None, xe=None,
```
Resoved in https://github.com/scipy/scipy/pull/23108

czgdp1807 commented on scipy/interpolate/src/__fitpack.cc:
```diff
@@ -262,6 +349,338 @@ qr_reduce(double *aptr, const int64_t m, const int64_t nz, // a(m, nz), packed
+    }
+}
+
+void _compute_residuals(
```
We don't call `_compute_residuals` in `qr_reduce`. We are calling it in `fpbacp` and `fpback`. `_get_residuals_p0` is entirely removed now.

czgdp1807 commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -39,10 +40,7 @@
 
-def _get_residuals(x, y, t, k, w):
-    # FITPACK has (w*(spl(x)-y))**2; make_lsq_spline has w*(spl(x)-y)**2
-    w2 = w**2
```
> Do we have test coverage with non-constant weights? For both periodic and non-periodic cases please.

Could you please explain what do you mean by non-constant weights? Like weights which are generated by a python function? Let me know.

czgdp1807 replied:
Added a test, `test_make_splrep_with_unequal_weights` in `scipy/interpolate/tests/test_bsplines.py`.

czgdp1807 commented:
So only one review is left to be addressed, https://github.com/czgdp1807/scipy/pull/19#discussion_r2104406503. However before addressing it I would recommend to review the PR again. If everything looks good then I would generalise for `y.ndim > 1`. 

cc: @ev-br @rgommers

czgdp1807 commented:
@ev-br I implemented the generalisation to `y.ndim > 1`. However it seems like the function which accepts such inputs is `make_splprep`. Do you want me to update for periodic splines? If so then I have a question related to its test case,

https://github.com/scipy/scipy/blob/d3d081d73748570fc15f16afed63068531cf4158/scipy/interpolate/tests/test_bsplines.py#L3596-L3597

Are the number of knots calculated theoretically? Which formula used for doing so?

ev-br commented:
@ev-br I implemented the generalisation to y.ndim > 1. 

That's great!

Yes, it's for `make_splprep`. The number of knots and the values of `s` in that test is just from trial-and-error: I wanted to make sure that the test probes various shortcuts and code paths (no internal knots, one iteration adding an internal knot, multiple iterations, a fall-back to `nmax` knots). I imagine that for periodic boundary condition, the numbers may change. So maybe add a similar test with appropriately adjusted values?

czgdp1807 commented:
All reviews addressed. This is fully ready for review. :-)

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -765,6 +1036,20 @@ def make_splrep(x, y, *, w=None, xb=None, xe=None, k=3, s=0, t=None, nest=None):
+        ``x[0]`` and a smooth periodic spline approximation is returned. Values of
+        ``y[m-1]`` and ``w[m-1]`` are not used.
+        The default is zero, corresponding to boundary condition 'not-a-knot'.
+    bc_type : 2-tuple or None
```
```suggestion
    bc_type : str, optional
```

ev-br commented on scipy/interpolate/_fitpack_repro.py:
```diff
@@ -163,6 +168,16 @@ def generate_knots(x, y, *, w=None, xb=None, xe=None, k=3, s=0, nest=None):
     nest : int, optional
         Stop when at least this many knots are placed.
+          ends are equivalent.
+    bc_type : 2-tuple or None
```
```suggestion
    bc_type : str, optional
```

ev-br reviewed (changes requested):
Also please update the `make_splprep` Parameters section.

ev-br commented on scipy/interpolate/src/_dierckxmodule.cc:
```diff
@@ -257,22 +592,54 @@ py_data_matrix(PyObject *self, PyObject *args)
+            // np.asarray(A), np.asarray(offset), int(nc)
+            PyObject *py_nc = PyLong_FromSsize_t(static_cast<Py_ssize_t>(nc));
+            return Py_BuildValue("(NNN)", PyArray_Return(a_A), PyArray_Return(a_offs), py_nc);
+        } else {
```
Could you please split this wrapper into two, separate `data_matrix_periodic`. AFAIU, the C++ implementations are separate, the python usage is different, so there's no reason to not have two separate functions.
